### DISCUSSION TITLE
Custom Laravel Package with Nova Resources & Laravel Models [help needed]

### DISCUSSION THREAD

**Discussion by NowakAdmin** [laravel/nova-issues#6558] [Q&A] [answered]
Hello everyone. 
I'm very fresh with managing composer/laravel packages and after my research i cannot accomplish my goal.

What I did:
I used this package to create shell for my own package:
`composer require jeroen-g/laravel-packager --dev`

created a shell package:
`php artisan packager:new NowakAdmin SomePackage`

Then I  added folders: 
Packages\NowakAdmin\SomePackage\src\Models <- there i putted my models files
Packages\NowakAdmin\SomePackage\src\Nova <- there i putted my nova resources files

Then i made a Github private repository of them, added to composer.json:
```
"require-dev": {
    "nowakadmin/somepackage": "dev-main",
}
...
"repositories": [
            {
                "type": "vcs",
                "url": "git@github.com:nowakadmin/somepackage.git"
            }

```
removed package from /Packages/:
`php artisan packager:remove NowakAdmin SomePackage`

Finally composer update - no errors, downloaded from github, installed.

But no models / resources loaded (not showing in nova menu)

I tried to do something simplier for beginning so I added code in 
```
    /**
     * Register any package services.
     *
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/namsbase.php', 'namsbase');

        // Nova Logo
        $this->setNovaLogo();

        // Nova footer
        Nova::footer(function ($request) {
            return Blade::render('
            <p class="mt-8 text-center text-xs text-80">
            <a href="" class="text-primary dim no-underline"></a>
            <span class="px-1">&middot;</span>
            &copy; {{ date("Y") }} by Nowak Admin
            <span class="px-1">&middot;</span>
            Nova {{ \Laravel\Nova\Nova::version() }} || PHP {{ phpversion() }} || Laravel {{ app()->version() }}
        </p>
            ', ['mysql' => $this->mysql_version()]);
        });
        
        // Register the service the package provides.
        $this->app->singleton('namsbase', function ($app) {
            return new NamsBase;
        });

    }
```
Above code work and overrides footer and logo of Nova. But idk if this is a good place to put this code to.

in composer.json whole src is loaded
```
"autoload": {
        "psr-4": {
            "NowakAdmin\\SomePackage\\": "src"
        }
    },
```

And in models i tried fewsolutions:
as in anwser post: https://stackoverflow.com/questions/19133020/using-models-on-packages
```
<?php
// Note no namespace

use \Illuminate\Database\Eloquent\Model as Eloquent;

class YourModel extends Eloquent {
    //
}
```

or as in https://www.laravelpackage.com/08-models-and-migrations/
```
<?php
namespace NowakAdmin\SomePackage\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class Distribution extends Model
{
```

The sources of above are somewaht old (2011/ 2018) and i assume that nowdays it should be done in other way.

Also i did not find anything about adding models to packages in Laravel docs: 
https://laravel.com/docs/11.x/packages

I assume that mby I just need to make some Nova Menu hook in my package? (it's in default, fresh-install mode)

I'm Using Laravel 11.27.2 and nova 4.35.3  on php 8.2.22; Hope my english is understandable.

To clarify:
My goal is to have laravel models connected with nova resources inside custom laravel package.

**Comment by ismaildasci** [comment, marked as answer]
Hey! 👋

You’re on the right track with creating a custom Laravel package and connecting it with Nova. Here’s a structured approach to ensure your models and Nova resources are properly recognized within your custom package:

### 1. Ensure the Package Service Provider is Registered

First, make sure your package’s service provider is registered. You can confirm this by checking that it’s loaded in your `composer.json` file (which you’ve done with `autoload`), and by using the `register` and `boot` methods in your service provider. 

In `src/SomePackageServiceProvider.php`, you should have something like this:

```php
namespace NowakAdmin\SomePackage;

use Illuminate\Support\ServiceProvider;
use Laravel\Nova\Nova;

class SomePackageServiceProvider extends ServiceProvider
{
    /**
     * Register any package services.
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/namsbase.php', 'namsbase');
        
        // Register the package's models or other bindings here.
        $this->app->singleton('namsbase', function ($app) {
            return new NamsBase;
        });
    }

    /**
     * Bootstrap any package services.
     */
    public function boot(): void
    {
        if ($this->app->runningInConsole()) {
            // Load migrations if you have them
            $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        }

        // Register Nova resources if they exist
        Nova::resources([
            \NowakAdmin\SomePackage\Nova\YourNovaResource::class,
            // Add other Nova resources as needed
        ]);

        // Load routes, views, etc. if applicable
        $this->loadRoutesFrom(__DIR__.'/../routes/web.php');
        $this->loadViewsFrom(__DIR__.'/../resources/views', 'somepackage');
    }
}
```

### 2. Configure the `composer.json` Autoload

Ensure your package is set up correctly in `composer.json`. It looks like you’ve already added the namespace in `autoload`. Don’t forget to run:

```bash
composer dump-autoload
```

This command reloads your package autoloading, which helps Laravel recognize new classes and files in your package.

### 3. Use Namespaces in Models and Nova Resources

Your models and Nova resources need to be properly namespaced to match the autoload configuration. Based on your structure, the models should have the following format:

```php
<?php

namespace NowakAdmin\SomePackage\Models;

use Illuminate\Database\Eloquent\Model;

class YourModel extends Model
{
    // Model code here
}
```

In the same way, your Nova resources should be in the `NowakAdmin\SomePackage\Nova` namespace:

```php
<?php

namespace NowakAdmin\SomePackage\Nova;

use Laravel\Nova\Resource;
use NowakAdmin\SomePackage\Models\YourModel;

class YourResource extends Resource
{
    public static $model = YourModel::class;

    // Define fields, cards, filters, etc.
}
```

### 4. Register Nova Resources Dynamically (Optional)

If you want to register all Nova resources in your package without manually listing each one, you can use the `Nova::resourcesIn` method in `boot`:

```php
public function boot(): void
{
    Nova::resourcesIn(__DIR__ . '/Nova');
}
```

### 5. Add Nova Menu and Hooks (Optional)

If you want your Nova resources to show in the Nova menu, ensure they are properly set up in the Nova resource class with the correct configuration. Nova should automatically add resources to the menu if they are registered, so make sure you have a correct Nova resource setup.

### 6. Confirm Nova’s Configuration for the Package

Run `php artisan nova:resource` to confirm Nova recognizes the resources. You should now see your package models listed in Nova.

### Example Structure

After following these steps, your package folder structure should look something like this:

```
Packages/
└── NowakAdmin/
    └── SomePackage/
        ├── src/
        │   ├── Models/
        │   │   └── YourModel.php
        │   ├── Nova/
        │   │   └── YourResource.php
        │   └── SomePackageServiceProvider.php
        ├── config/
        │   └── namsbase.php
        ├── routes/
        │   └── web.php
        └── composer.json
```

### Summary Checklist

1. Confirm `SomePackageServiceProvider` is loaded.
2. Verify that models and Nova resources are namespaced correctly.
3. Register Nova resources in `SomePackageServiceProvider`.
4. Check the Nova resources in the Laravel app after `composer dump-autoload`.

Let me know if this works for you or if there are any specific issues! 😊
